Archive/ZagierTwoSquares.lean

```diff
@@ -0,0 +1,183 @@
+  unfold obvInvo at this; simp at this
+  unfold zagierSet at he; simp at he
+  use x, (2 * y)
+  rw [show 2 * y * (2 * y) = 4 * y * y by linarith, ← he, this.1]
+
+/-- The complicated involution, which is shown to have exactly one fixed point `(1, 1, k)`. -/
+def complexInvo : Function.End (zagierSet k) := fun ⟨⟨x, y, z⟩, h⟩ =>
+  ⟨if x + z < y then ⟨x + 2 * z, z, y - x - z⟩ else
+   if 2 * y < x then ⟨x - 2 * y, x + z - y, y⟩ else
+                     ⟨2 * y - x, y, x + z - y⟩, by
+  unfold zagierSet at h ⊢; simp at h
+  split_ifs with less more <;> simp
+  · rw [Nat.sub_sub]; zify [less.le] at h ⊢; linarith [h]
+  · push_neg at less; zify [less, more.le] at h ⊢; linarith [h]
+  · push_neg at less more; zify [less, more] at h ⊢; linarith [h]⟩
+
+theorem complexInvo_sq : complexInvo k ^ 2 = 1 := by
+  change complexInvo k ∘ complexInvo k = id
+  funext ⟨⟨x, y, z⟩, h⟩
+  rw [comp_apply]
+  obtain ⟨xb, _, _⟩ := zagierSet_lower_bound k h
+  conv_lhs => arg 2; tactic => unfold complexInvo; dsimp
+  split_ifs with less more <;> (unfold complexInvo; dsimp)
+  · simp [show ¬(x + 2 * z + (y - x - z) < z) by linarith [less], xb]
+    rw [Nat.sub_sub, two_mul, ← tsub_add_eq_add_tsub (by linarith), ← add_assoc,
+      Nat.add_sub_cancel, add_comm (x + z), Nat.sub_add_cancel (less.le)]
+  · push_neg at less
+    simp [(show x - 2 * y + y < x + z - y by zify [more.le, less]; linarith)]
+    constructor
+    · rw [Nat.sub_add_cancel more.le]
+    · rw [Nat.sub_right_comm, Nat.sub_sub _ _ y, ← two_mul, add_comm, Nat.add_sub_assoc more.le,
+        Nat.add_sub_cancel]
+  · push_neg at less more
+    simp [(show ¬(2 * y - x + (x + z - y) < y) by push_neg; zify [less, more]; linarith),
+      (show ¬(2 * y < 2 * y - x) by push_neg; zify [more]; linarith)]
+    constructor
+    · rw [tsub_tsub_assoc (by rfl) more, tsub_self, zero_add]
+    · rw [← Nat.add_sub_assoc less (2 * y - x), ← add_assoc, Nat.sub_add_cancel more,
+        Nat.sub_sub _ _ y, ← two_mul, add_comm, Nat.add_sub_cancel]
+
+theorem unique_of_mem_fixedPoints {t : zagierSet k} (mem : t ∈ fixedPoints (complexInvo k)) :
+    t.val = (1, 1, k) := by
+  rw [mem_fixedPoints_iff] at mem
+  unfold complexInvo at mem
+  obtain ⟨⟨x, y, z⟩, h⟩ := t
+  obtain ⟨xb, yb, zb⟩ := zagierSet_lower_bound k h
+  simp at mem
+  split_ifs at mem with less more <;> simp_all
+  · obtain ⟨_, _, _⟩ := mem; simp_all
+  · unfold zagierSet at h; simp at h
+    replace mem := mem.1
+    rw [tsub_eq_iff_eq_add_of_le more, ← two_mul] at mem
+    replace mem := (mul_left_cancel₀ two_ne_zero mem).symm
+    subst mem
+    rw [show x * x + 4 * x * z = x * (x + 4 * z) by linarith] at h
+    cases' (Nat.dvd_prime hk.out).1 (dvd_of_mul_left_eq _ h) with e e
+    · rw [e, mul_one] at h; subst h; simp_all [show z = 0 by linarith [e]]
+    · rw [e, mul_left_eq_self₀] at h; simp_all; linarith [e]
+
+theorem fixedPoints_eq_singleton :
+    fixedPoints (complexInvo k) = {⟨(1, 1, k), (by unfold zagierSet; simp; linarith)⟩} := by
```

Comment:
I think the RHS should have its own definition, using `simp` and `linarith` in the statement of a theorem looks not great.

Reply:
All done (including squeezing nonterminal simps and documentation).